File: Structural/Bridge/DIY.java
package Structural.Bridge;

public class DIY {
    public static void main(String[] args) {
    //Can be separated
        var remoteControl = new BasicRemoteControl(new TV());
        testDevice(new TV());
        remoteControl.changeChannel();
        testDevice(new Radio());
    }
    //Can be put together
    public static void testDevice(Device device){
        System.out.println("Testing with the Basic Remote Control");
        BasicRemoteControl basicRemote = new BasicRemoteControl(device);
        basicRemote.power();
        device.printStatus();

        System.out.println("Testing with the Advanced Remote Control");
        AdvancedRemoteControl advanceRemote = new AdvancedRemoteControl(device);
        advanceRemote.power();
        advanceRemote.mute();
        device.printStatus();
    }

    
}

class BasicRemoteControl{
protected Device device;

public BasicRemoteControl(Device device) {
    this.device = device;
}

  void power(){
    System.out.println("Remote: Device is toggled on.");
    if(device.isEnabled()){
        device.disable();
      
    }
    device.enable();
  }

  void volumeDown(){
  
    System.out.println("Lowering volume...");
  }

  void volumeUp(){
    System.out.println("Extending volume...");
  }
  void changeChannel(){
    System.out.println("Changing channel...");
  }


}

class AdvancedRemoteControl extends BasicRemoteControl{
    public AdvancedRemoteControl(Device device){
        super(device);
    }
    public void mute(){
        System.out.println("Volume Muted");
    }
}

interface Device{
    boolean isEnabled();
    
    void enable();

    void disable();

    int getVolume();

    void setVolume(int percent);

    int getChannel();

    void setChannel(double channel);

    void printStatus();

}

class TV implements Device{
    private boolean on = false;
    private int volume= 30;
    private double channel=7;

    
    @Override
    public int getVolume() {
        return volume;
    }

    @Override
    public void setVolume(int volume) {
            this.volume=volume;
    }

    @Override
    public int getChannel() {
        return (int)channel;
    }

    @Override
    public void setChannel(double channel) {
      this.channel = Math.round(channel);
    }

    @Override
    public void printStatus() {
        System.out.println("------------------------------------");
        System.out.println("| Model: TV-Panasonic W0ODQ123");
        System.out.println("| Status: " + (on? "enabled" : "disabled"));
        System.out.println("| Volume level: " + getVolume() + "%");
        System.out.println("| Current channel is " + channel);
        System.out.println("------------------------------------\n");
    }

    @Override
    public boolean isEnabled() {
        return on;
    }

    @Override
    public void enable() {
       on = true;
    }

    @Override
    public void disable() {
       on = false;
    }

    

}

class Radio implements Device{
    private boolean on = false;
    private int volume=20;
    private double channel=91.9;

    @Override
    public boolean isEnabled() {
        return on;
        
    }

    @Override
    public void enable() {
        on = true;
    }

    @Override
    public void disable() {
        on = false;
        
    }

    @Override
    public int getVolume() {
        return volume;
    }

    @Override
    public void setVolume(int percent) {
       this.volume = volume;
    }

    @Override
    public int getChannel() {
       return (int)channel;
    }

    @Override
    public void setChannel(double channel) {
       this.channel = channel;
    }

    @Override
    public void printStatus() {
        System.out.println("------------------------------------");
        System.out.println("| Model: RADIO1XQ45");
        System.out.println("| Status:  " + (on? "enabled" : "disabled"));
        System.out.println("| Volume level: " + getVolume() + "%");
        System.out.println("| Current channel is " + channel);
        System.out.println("------------------------------------\n");
    }
}

File: TS/Structural/Bridge/DIY.ts
interface Device {
  isEnabled(): boolean;
  enable(): void;
  disable(): void;
  getVolume(): number;
  setVolume(percent: number): void;
  getChannel(): number;
  setChannel(channel: number): void;
  printStatus(): void;
}

class TV implements Device {
  private on: boolean = false;
  private volume: number = 30;
  private channel: number = 7;

  isEnabled(): boolean {
    return this.on;
  }

  enable(): void {
    this.on = true;
  }

  disable(): void {
    this.on = false;
  }

  getVolume(): number {
    return this.volume;
  }

  setVolume(volume: number): void {
    this.volume = volume;
  }

  getChannel(): number {
    return this.channel;
  }

  setChannel(channel: number): void {
    this.channel = Math.round(channel);
  }

  printStatus(): void {
    console.log("------------------------------------");
    console.log("| Model: TV-Panasonic W0ODQ123");
    console.log("| Status: " + (this.on ? "enabled" : "disabled"));
    console.log("| Volume level: " + this.getVolume() + "%");
    console.log("| Current channel is " + this.channel);
    console.log("------------------------------------\n");
  }
}

class Radio implements Device {
  private on: boolean = false;
  private volume: number = 20;
  private channel: number = 91.9;

  isEnabled(): boolean {
    return this.on;
  }

  enable(): void {
    this.on = true;
  }

  disable(): void {
    this.on = false;
  }

  getVolume(): number {
    return this.volume;
  }

  setVolume(volume: number): void {
    this.volume = volume;
  }

  getChannel(): number {
    return this.channel;
  }

  setChannel(channel: number): void {
    this.channel = channel;
  }

  printStatus(): void {
    console.log("------------------------------------");
    console.log("| Model: RADIO1XQ45");
    console.log("| Status:  " + (this.on ? "enabled" : "disabled"));
    console.log("| Volume level: " + this.getVolume() + "%");
    console.log("| Current channel is " + this.channel);
    console.log("------------------------------------\n");
  }
}

class BasicRemoteControl {
  protected device: Device;

  constructor(device: Device) {
    this.device = device;
  }

  power(): void {
    console.log("Remote: Device is toggled on.");
    if (this.device.isEnabled()) {
      this.device.disable();
    } else {
      this.device.enable();
    }
  }

  volumeDown(): void {
    console.log("Lowering volume...");
  }

  volumeUp(): void {
    console.log("Extending volume...");
  }

  changeChannel(): void {
    console.log("Changing channel...");
  }
}

class AdvancedRemoteControl extends BasicRemoteControl {
  constructor(device: Device) {
    super(device);
  }

  mute(): void {
    console.log("Volume Muted");
  }
}

class DIYBridge {
  public static main(): void {
    // Can be separated
    const remoteControl = new BasicRemoteControl(new TV());
    DIYBridge.testDevice(new TV());
    remoteControl.changeChannel();
    DIYBridge.testDevice(new Radio());
  }

  // Can be put together
  public static testDevice(device: Device): void {
    console.log("Testing with the Basic Remote Control");
    const basicRemote = new BasicRemoteControl(device);
    basicRemote.power();
    device.printStatus();

    console.log("Testing with the Advanced Remote Control");
    const advanceRemote = new AdvancedRemoteControl(device);
    advanceRemote.power();
    advanceRemote.mute();
    device.printStatus();
  }
}

DIYBridge.main();
